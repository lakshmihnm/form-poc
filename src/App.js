import "./App.css";
import "../node_modules/bootstrap/dist/css/bootstrap.min.css";
import React, { useState, useEffect } from "react";
import styled from "styled-components";
import validate from "./components/formValidations";
import useForm from "./components/useForm";

const Form = styled.form`
  display: flex;
  flex-direction: column;
  width: 400px;
  min-width: 100px;
  padding: 20px 40px 40px 40px;
  border-radius: 6px;
  box-shadow: 0px 8px 36px #222;
  background-color: #fefefe;
`;
const SuccessDiv = styled.header`
  display: flex;
  flex-direction: column;
  width: 400px;
  min-width: 100px;
  height: opx;
  padding: 20px 40px 40px 40px;
  border-radius: 6px;
  box-shadow: 0px 8px 36px #222;
  background-color: #fefefe;
  margin-bottom: 10px;
  border-top: 3px solid green;
`;
const Div = styled.div`
  position: absolute;
  top: 50%;
  left: 35%;
  transform: translate(0, -50%);
  // border: 5px solid #FFFF00;
  padding: 10px;
`;

const Header = styled.h3`
  display: flex;
  margin-bottom: 20px;
`;
const Small = styled.small`
  color: #db2269;
  font-size: 0.8em;
  display: relative;
`;

function App() {
  const [flag, setFlag] = useState(false);
  const formData = [
    {
      type: "text",
      id: "labelName",
      name: "name",
      placeholder: "Enter Name",
      ariaLabel: "Please enter name",
      label: "Name",
    },
    {
      type: "number",
      id: "labelNumber",
      name: "phone",
      placeholder: "Enter contact Number",
      ariaLabel: "Please enter contact number",
      label: "Contact",
    },
    {
      type: "date",
      id: "labelDate",
      name: "doj",
      placeholder: "Enter DOJ",
      ariaLabel: "Please enter DOJ",
      label: "DOJ",
    },
  ];

  const handleSubmision = () => {
    console.log("No errors");
    setTimeout(() => {
      setFlag(true);
    }, 2000);
  };

  useEffect(() => {
    return () => {
      setTimeout(() => {
        setFlag(false);
      }, 2000);
    };
  });

  const { values, errors, handleChange, handleSubmit } = useForm(
    handleSubmision,
    validate
  );
  return (
    <Div>
      <Header>Employee Details Form</Header>
      {flag && <SuccessDiv>Form Submitted Successfully</SuccessDiv>}
      <Form>
        {/* <h3>Details</h3> */}

        {formData.map((data) => {
          const { name, id, ariaLabel, placeholder, type, label } = data;
          return (
            <div className="form-group" key={id}>
              <label htmlFor={name} id={id} aria-label={ariaLabel}>
                {label}
              </label>
              <input
                type={type}
                className="form-control"
                name={name}
                id={name}
                placeholder={placeholder}
                onChange={handleChange}
                aria-labelledby={id}
                data-testid={id}
                required
              />
              {errors[name] && (
                <p>
                  <Small data-testid={errors[name]}>{errors[name]}</Small>
                </p>
              )}
            </div>
          );
        })}

        <button
          type="submit"
          onClick={handleSubmit}
          className="btn btn-dark btn-lg btn-block"
        >
          Submit
        </button>
      </Form>
    </Div>
  );
}

export default App;
