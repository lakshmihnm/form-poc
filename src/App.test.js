import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import App from "./App";
import { renderer, cleanup } from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event";

// afterEach(cleanup);

// it("matches snapshot", () => {
//   const { asFragment } = render(<App />);
//   expect(asFragment()).toMatchSnapshot();
// });

test("renders Employee Details Form header", () => {
  render(<App />);
  const linkElement = screen.getByText(/Employee Details Form/i);
  expect(linkElement).toBeInTheDocument();
});

test("renders the form correctly", () => {
  const { getByText, getByLabelText } = render(<App />);
  const nameLabel = getByLabelText("Name");
  const contactLabel = getByLabelText("Contact");
  const dojLabel = getByLabelText("DOJ");
  const button = getByText(/Submit/i);
  expect(button).toBeInTheDocument();
  expect(nameLabel).toBeInTheDocument();
  expect(contactLabel).toBeInTheDocument();
  expect(dojLabel).toBeInTheDocument();
  expect(nameLabel).toHaveAttribute("type", "text");
  expect(contactLabel).toHaveAttribute("type", "number");
  expect(dojLabel).toHaveAttribute("type", "date");
});

test("pass value to test name input field", () => {
  const { getByTestId, queryByTestId } = render(<App />);

  const inputEl = getByTestId("labelName");
  userEvent.type(inputEl, "test");

  expect(getByTestId("labelName")).toHaveValue("test");
  expect(queryByTestId("name")).not.toBeInTheDocument();
});

test("submit button should be disabled when Name is empty", () => {
  const { getByLabelText, getByRole, debug } = render(<App />);
  const input = getByLabelText("Name");
  fireEvent.change(input, { target: { value: "" } });
  const submitBtn = getByRole("button", { name: "Submit" });
  expect(submitBtn).not.toHaveAttribute({ disabled: false });
  debug(submitBtn);
  fireEvent.change(input, { target: { value: "John Doe" } });
  debug(submitBtn);
  expect(submitBtn).not.toHaveAttribute({ disabled: true });
});
