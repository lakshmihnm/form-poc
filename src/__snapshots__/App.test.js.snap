// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-dkPtyc dZVVTd"
  >
    <h3
      class="sc-hKwCoD hjtCOr"
    >
      Employee Details Form
    </h3>
    <form
      class="sc-bdvvaa bvxtZM"
    >
      <div
        class="form-group"
      >
        <label
          aria-label="Please enter name"
          for="name"
          id="labelName"
        >
          Name
        </label>
        <input
          aria-labelledby="labelName"
          class="form-control"
          data-testid="labelName"
          id="name"
          name="name"
          placeholder="Enter Name"
          required=""
          type="text"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          aria-label="Please enter contact number"
          for="phone"
          id="labelNumber"
        >
          Contact
        </label>
        <input
          aria-labelledby="labelNumber"
          class="form-control"
          data-testid="labelNumber"
          id="phone"
          name="phone"
          placeholder="Enter contact Number"
          required=""
          type="number"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          aria-label="Please enter DOJ"
          for="doj"
          id="labelDate"
        >
          DOJ
        </label>
        <input
          aria-labelledby="labelDate"
          class="form-control"
          data-testid="labelDate"
          id="doj"
          name="doj"
          placeholder="Enter DOJ"
          required=""
          type="date"
        />
      </div>
      <button
        class="btn btn-dark btn-lg btn-block"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
